### DISCUSSION TITLE
Block where a specific object instance is implied when accessing instance members

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#7482] [Language Ideas]
## Summary

Allow accessing instance members (properties, events, methods) on a specific instance without using the name of the instance before the access, by designating it inside a block statement for this purpose.

## Design

A new block-encapsulating statement will be introduced allowing accessing those members on the instance that is given. The syntax will look like this:
```csharp
on instanceExpression
{
    SomeMethod();
    SomeOtherMethod();
    Property = 3;
}
```

The above code snippet is equivalent to this:
```csharp
// potentially needing a temporary
var instanceExpression = ...;
instanceExpression.SomeMethod();
instanceExpression.SomeOtherMethod();
instanceExpression.Property = 3;
```

The statement can contain multiple nested statements, binding the qualified members to the innermost instance that contains a member with the target name. This also applies for the current `this` instance; if the name cannot be found on the instance in the `on` statement, the member access will refer to the `this` instance. For example:

Consider
- `this` with the methods `T()`, `U()`,
- `outerInstance` with the property `P` and
- `innerInstance` with the property `Q`:

```csharp
on outerInstance
{
    T(); // this.T();
    P = 3; // outerInstance.P = 3;
    on innerInstance
    {
        T(); // this.T();
        Q = 3; // innerInstance.P = 3;
        P = 3; // outerInstance.P = 3;
    }
}
```

The `this` qualifier bypasses the block's instance, and resorts to the actual `this` instance; the instance of the type definition. With the above definitions:
```csharp
on outerInstance
{
    this.T();
}
```
resorts to `this`, and not `outerInstance`. If `outerInstance` contained a `T()` method, the `this` instance would still be used, escaping the binding to `outerInstance` without the qualifications.

Normal qualifications are still generally available, even when including the instance in the block.

## Effect

This helps applying methods and setting properties on existing instances outside the declaration of the instance, reducing copy-paste of the same instance identifier.

## Open Questions

- Binding of conflicting method names

**Comment by iam3yal** [comment]
@Rekkonnect Many times this sort of setup comes up when constructing objects, factories, fluent APIs and DSLs, I'd also want them to enhance object initializers to allow for more stuff there so I think that there should be independent features that compose well together to address these scenarios and my guess that this goes beyond just syntax so introducing new syntax just to address a very specific problem would be a shame imo.

e.g., It would have been awesome if the language could recognize factory methods so we had the ability to do something like this:
```csharp
X.Create() {
 Click += (_, _) => Console.WriteLine("Hey!"),
 Foo = "F",
 Bar()
}
```
Similar to the way object initializer work.

**Reply by Rekkonnect** [reply]
The proposal is not just about further initializing instances from factory methods or whatever instance creation source. You get any expression available, you throw it into the `on` block, and you now treat it as your main instance on which you use members. Enhanced initializers is slightly out of scope of this proposal.

**Reply by iam3yal** [reply]
@Rekkonnect Here is the thing enhanced initializers should not be out of scope for this because I doubt we want to have various ways to achieve all the things I mentioned so if you want to do X you have to do Y and if you want to do N you have to do M this can lead to friction and frustration when we can have a design that may cover all of these through maybe independent features over multiple releases but the big picture definitely needs to be there first.

**Reply by Rekkonnect** [reply]
They are out of scope because initializers cannot be used beyond the initialization of an object, whereas this proposal is applied to any stage of the object's lifetime, primarily post-initialization.

Nevertheless, the primary goal is to reduce repetition when filling an object's properties, or performing certain tasks. The initialization part can also get enhancements, and have behavior align more towards "instance scoping".

**Reply by iam3yal** [reply]
@Rekkonnect Yes I know, what I'm trying to explain to you is that there shouldn't be a syntax for X and syntax for Y there should be a syntax that can do both or improve upon existing syntax and not just for the sake of reducing repetitions.

**Reply by Rekkonnect** [reply]
Yes, I was planning for this proposal to also offer extensibility on new object initialization, or even factory methods. If you notice, I'm saying "instance **expression**", which extends to assignments, method invocations and alike.

While variable declarations aren't expressions, but statements, I'd like the idea of extending that.

So for example, you'd have
```csharp
on var options = CreateOptions(1)
{
    ExtraParameter = new object();
}
```

Notice the only difference being the semicolon instead of the comma, even when assigning properties or fields.

**Reply by HaloFour** [reply]
I think the conversation always moved to object initializers because that's what the members of the team had any interest in doing.  For all of the times that `with` has been requested in the language over the past 20 years (and it's a shame that the blog post from 2004 doesn't appear to be saved in the Wayback Machine), different members of the language team didn't find eliminating an assignment to a temporary variable worthy of a language feature.

**Comment by HaloFour** [comment]
See: #6602

Many VB/Pascal/Javascript `with` block proposals have been linked off of that discussion as well.
